src/meta-srv/src/election/mysql.rs

```diff
@@ -0,0 +1,1553 @@
+use common_telemetry::{error, info, warn};
+use common_time::Timestamp;
+use itertools::Itertools;
+use snafu::{ensure, OptionExt, ResultExt};
+use sqlx::mysql::{MySqlArguments, MySqlRow};
+use sqlx::query::Query;
+use sqlx::{MySql, MySqlConnection, MySqlTransaction, Row};
+use tokio::sync::{broadcast, Mutex, MutexGuard};
+use tokio::time::{Interval, MissedTickBehavior};
+
+use crate::election::{
+    listen_leader_change, Election, LeaderChangeMessage, LeaderKey, CANDIDATES_ROOT, ELECTION_KEY,
+    IDLE_SESSION_TIMEOUT,
+};
+use crate::error::{
+    DeserializeFromJsonSnafu, MySqlExecutionSnafu, NoLeaderSnafu, Result, SerializeToJsonSnafu,
+    UnexpectedSnafu,
+};
+use crate::metasrv::{ElectionRef, LeaderValue, MetasrvNodeInfo};
+
+// Separator between value and expire time.
+const LEASE_SEP: &str = r#"||__metadata_lease_sep||"#;
+
+struct ElectionSqlFactory<'a> {
+    table_name: &'a str,
+}
+
+struct ElectionSqlSet {
+    campaign: String,
+    // SQL to put a value with expire time.
+    //
+    // Parameters for the query:
+    // `$1`: key,
+    // `$2`: value,
+    // `$3`: lease time in seconds
+    //
+    // Returns:
+    // If the key already exists, return the previous value.
+    put_value_with_lease: String,
+    // SQL to update a value with expire time.
+    //
+    // Parameters for the query:
+    // `$1`: updated value,
+    // `$2`: lease time in seconds
+    // `$3`: key,
+    // `$4`: previous value,
+    update_value_with_lease: String,
+    // SQL to get a value with expire time.
+    //
+    // Parameters:
+    // `$1`: key
+    get_value_with_lease: String,
+    // SQL to get all values with expire time with the given key prefix.
+    //
+    // Parameters:
+    // `$1`: key prefix like 'prefix%'
+    //
+    // Returns:
+    // column 0: value,
+    // column 1: current timestamp
+    get_value_with_lease_by_prefix: String,
+    // SQL to delete a value.
+    //
+    // Parameters:
+    // `?`: key
+    //
+    // Returns:
+    // Rows affected
+    delete_value: String,
+}
+
+impl<'a> ElectionSqlFactory<'a> {
+    fn new(table_name: &'a str) -> Self {
+        Self { table_name }
+    }
+
+    fn build(self) -> ElectionSqlSet {
+        ElectionSqlSet {
+            campaign: self.campaign_sql(),
+            put_value_with_lease: self.put_value_with_lease_sql(),
+            update_value_with_lease: self.update_value_with_lease_sql(),
+            get_value_with_lease: self.get_value_with_lease_sql(),
+            get_value_with_lease_by_prefix: self.get_value_with_lease_by_prefix_sql(),
+            delete_value: self.delete_value_sql(),
+        }
+    }
+
+    // Currently the session timeout is longer than the leader lease time, so the leader lease may expire while the session is still alive.
+    // Either the leader reconnects and step down or the session expires and the lock is released.
+    fn set_idle_session_timeout_sql(&self) -> String {
```

Comment:
If session timeout has a connection with leader lease timeout, we should pass the leader lease timeout into this function and calculate the session timeout out of it. I think the session timeout should be always longer than leader lease timeout, because we don't want session expiry interferes with leader election.

Reply:
Now I set it to `META_LEASE_SECS + 1`. The client will renew the lease twice if everything goes well.